crates/sim2h/src/lib.rs

```diff
@@ -684,128 +746,330 @@ impl Sim2hState {
-            "sim2h-priv_check_incoming_connections",
-            self.metric_publisher,
-            || {
-                if let Ok(wss) = self.wss_recv.try_recv() {
+        let _m = self
+            .metric_gen
+            .timer("sim2h-priv_check_incoming_connections");
+
+        let mut did_work = false;
+        let mut wss_list = Vec::new();
+        for _ in 0..100 {
+            if let Ok(wss) = self.wss_recv.try_recv() {
+                did_work = true;
+                wss_list.push(wss);
+            } else {
+                break;
+            }
+        }
+        if !wss_list.is_empty() {
+            //let job_send = self.pool.get_push_job_handle();
+            let msg_send = self.msg_send.clone();
+            let sim2h_handle = self.sim2h_handle.clone();
+            tokio::task::spawn(async move {
+                let _m =
+                    sim2h_handle.metric_timer("sim2h-priv_check_incoming_connections-async-add");
+                let mut state = sim2h_handle.lock_state().await;
+
+                for wss in wss_list.drain(..) {
                     let url: Lib3hUri = url::Url::from(wss.remote_url()).into();
-                    let (job, outgoing_send) = ConnectionJob::new(wss, self.msg_send.clone());
-                    let job = Arc::new(Mutex::new(job));
-                    if let Err(error) = self.handle_incoming_connect(url.clone()) {
-                        error!("Error handling incoming connection: {:?}", error);
-                        return true; //did work despite error.
-                    }
                     let uuid = nanoid::simple();
                     open_lifecycle("adding conn job", &uuid, &url);
-                    self.state.write().open_connections.insert(
+
+                    let (job, outgoing_send) = ConnectionJob::new(wss, msg_send.clone());
+                    let job = Arc::new(Mutex::new(job));
+
+                    //job_send.send(Box::new(job.clone())).expect("send fail");
+                    let mut job_clone = job.clone();
+                    tokio::task::spawn(async move {
+                        loop {
+                            let res = job_clone.run();
+                            if !res.cont {
+                                break;
+                            }
+                            if res.wait_ms == 0 {
+                                tokio::task::yield_now().await;
+                            } else {
+                                tokio::time::delay_for(std::time::Duration::from_millis(
+                                    res.wait_ms,
+                                ))
+                                .await;
+                            }
+                        }
+                    });
+
+                    state
+                        .connection_states
+                        .insert(url.clone(), (nanoid::simple(), ConnectionState::new()));
+
+                    state.open_connections.insert(
```

Comment:
Is there a possibility that spawning the Job into a tokio task before it's been added the `open_connections` list could cause a problem?  I.e. could it complete and try to remove itself from the open_connections list before it was added, or could it require a check on the open_connections list also before it's been added?

Reply:
The only thing the job does as a separate task (this was true before this PR as well) is publish messages and errors from reading / writing on the socket. Those messages don't get polled until another call to process happens...

I don't *think* it should make any difference... but it also shouldn't hurt to swap the ordering here, so I'll go ahead and do that.